master/internal/db/postgres_trial.go

```diff
@@ -468,99 +468,133 @@ func (db *PgDB) updateLatestValidationID(ctx context.Context, tx *sqlx.Tx, trial
-			SELECT max(q.total_batches) AS new_max_total_batches_processed
-			FROM (
-			SELECT coalesce(max(s.total_batches), 0) AS total_batches
-			FROM steps s
-			WHERE s.trial_id = $1
-			UNION ALL
-			SELECT coalesce(max(v.total_batches), 0) AS total_batches
-			FROM validations v
-			WHERE v.trial_id = $1
-		) q
-		) AS sub;
+				SELECT max(m.total_batches) AS total_batches_processed
+				FROM metrics m
+				WHERE m.trial_id = $1 AND m.archived = false
+			) AS latest;
 		`, trialID); err != nil {
 		return errors.Wrap(err, "error computing total_batches")
 	}
 	return nil
 }
 
+/*
+rollbackMetrics ensures old training and validation metrics from a previous run id are archived.
+DISCUSS: how do we decide if a utility should be namespaced under PgDB or not?
+the goal is to make it clear we don't have direct db access other than through tx.
+*/
+func rollbackMetrics(ctx context.Context, tx *sqlx.Tx, runID, trialID,
+	lastProcessedBatch int32, isValidation bool,
+) (int, error) {
+	pType := model.TrainingMetric
+	if isValidation {
+		pType = model.ValidationMetric
+	}
+
+	res, err := tx.ExecContext(ctx, `
+UPDATE metrics SET archived = true
+WHERE trial_id = $1
+  AND archived = false
+  AND trial_run_id < $2
+	-- we mark metrics reported in the same table with the same batch number
+	-- as the metric being added as "archived"
+  AND (
+		(
+			partition_type != $4 AND total_batches > $3
+		) OR
+		(
+			partition_type = $4 AND total_batches >= $3
+		)
+		
+	);
+	`, trialID, runID, lastProcessedBatch, pType)
+	if err != nil {
+		return 0, errors.Wrap(err, "archiving metrics")
+	}
+	affectedRows, err := res.RowsAffected()
+	if err != nil {
+		return 0, errors.Wrap(err, "checking for metric rollbacks")
+	}
+	return int(affectedRows), nil
+}
+
+func (db *PgDB) addRawMetrics(ctx context.Context, tx *sqlx.Tx, metricsBody *map[string]interface{},
+	runID, trialID, lastProcessedBatch int32,
+	pType model.MetricPartitionType, mType *string,
+) (int, error) {
+	var metricRowID int
+	if err := tx.QueryRowContext(ctx, `
+INSERT INTO metrics
+	(trial_id, trial_run_id, end_time, metrics, total_batches, partition_type, custom_type)
+VALUES
+	($1, $2, now(), $3, $4, $5, $6)
+RETURNING id`,
+		trialID, runID, *metricsBody, lastProcessedBatch, pType, mType, // CHECK: are nulls handled?
+	).Scan(&metricRowID); err != nil {
+		return metricRowID, errors.Wrap(err, "inserting metrics")
+	}
+
+	return metricRowID, nil
+}
+
 // AddTrialMetrics inserts a set of trial metrics to the database.
 func (db *PgDB) addTrialMetrics(
-	ctx context.Context, m *trialv1.TrialMetrics, isValidation bool,
-) (rollbacks map[string]int, err error) {
-	rollbacks = make(map[string]int)
-	trialMetricTables := []string{"raw_steps", "raw_validations"}
-	targetTable := "raw_steps"
+	ctx context.Context, m *trialv1.TrialMetrics, pType model.MetricPartitionType,
+	mType *string,
+) (rollbacks int, err error) {
+	/*
+		TODO(hamid):
+		- MetricType could live in trialv1.TrialMetrics :thinking_face:
```

Comment:
yep, I am not opposed to it being an `optional string ...`

Reply:
my only concern was that it could lead to a lot more data on the wire carrying duplicate data depending on when we set it